Add ncurses checker

param211 opened the pull request:
Fixes #116

param211 commented:
Steps followed:
- download ncurses https://ftp.gnu.org/pub/gnu/ncurses/
- extract
- ./configure --prefix=/opt/ncurses
- make
- sudo make install

Following strings were found:
**6.2**
![Screenshot from 2020-03-08 16-06-04](https://user-images.githubusercontent.com/20379896/76171412-2240d000-61b1-11ea-88c0-deab1924d522.png)

**6.1**
![Screenshot from 2020-03-08 16-21-58](https://user-images.githubusercontent.com/20379896/76171417-34227300-61b1-11ea-8565-b9857baa098e.png)

**5.8**
![Screenshot from 2020-03-08 16-52-26](https://user-images.githubusercontent.com/20379896/76171421-40a6cb80-61b1-11ea-8671-f971d519f66e.png)

**regex**: "ncurses ([0-9]+\.[0-9]+)"

Note: version before 6.1 require [patch](https://stackoverflow.com/questions/37475222/ncurses-6-0-compilation-error-error-expected-before-int) before installation

For "contains":
"GNU Screen" uses ncurses ([list here](https://www.openhub.net/tags/ncurses))
![Screenshot from 2020-03-08 19-22-53](https://user-images.githubusercontent.com/20379896/76171455-c296f480-61b1-11ea-9651-1243a83733cc.png)
-------------------------

Found installed libraries from here: http://www.linuxfromscratch.org/lfs/view/development/chapter06/ncurses.html

Possible mistakes/errors :
- selected wrong file for "test against real binary file"
-  using wrong string to check for "contains"

terriko commented on cve_bin_tool/checkers/ncurses.py:
```diff
@@ -0,0 +1,62 @@
+    """Tries to determine if a file includes Ncurses
+    """
+    for line in lines:
+        if "NCURSES" in line:
```
I'm not convinced this is the most unique string, since it seems like it might occur even if NCURSES was dynamically linked.  I don't suppose there were any better ones available?

terriko commented on test/binaries/test-ncurses-6.1.c:
```diff
@@ -0,0 +1,12 @@
+	printf("They appear below this line.");
+	printf("------------------");
+	printf("NCURSES");
+	printf("ncurses 6.1");
```
Could you change this to be a version that doesn't match the real file you're testing below?  (So basically anything other than 6.1 here, or change the test below to be somethign other than 6.1)  We've got a bug that's causing the one test result to mess up the other in test_scanner and I want to be 100% sure that this is working as intended.

terriko commented on test/test_scanner.py:
```diff
@@ -678,6 +689,12 @@ def test_binaries(self, binary, package, version, are_in, not_in):
                     ),
+                    (
+                        "https://ftp.gnu.org/pub/gnu/ncurses/",
+                        "ncurses-6.1.tar.gz",
```
Are there actually compiled ncurses binaries in this file?  Usually the gnu.org downloads are source only.  You may need to use a pre-compiled package from elsewhere to get this test to work.

param211 commented:
@terriko I made the changes you suggested. The binary file is for 6.1 and the real file is for 6.2. Also updated the contains string to something that is more unique and added a second regex (the first one doesnt seem to be present in fedora's pkg)

terriko commented:
Does the original regex weork against something like screen that you're pretty sure contains ncurses?

param211 commented:
> Does the original regex weork against something like screen that you're pretty sure contains ncurses?

The screen package from sources like rpmfind doesn't contain any string relating it to ncurses. The local screen binary file (which I got after compiling screen from its source) contains the string "NCURSES". 

Ncurses' [wikipedia page](https://en.wikipedia.org/wiki/Ncurses) mentions the programs using ncurse (although it looks like in a stripped-down manner, possibly explaining why "ncurses" isn't present in rpmfind's version but present in a locally compiled version).

terriko commented:
It's also possible that screen uses the dynamic library of ncurses, in which case it wouldn't contain any of the strings (and cve-bin-tool shouldn't detect it as containing ncurses)

terriko commented on cve_bin_tool/checkers/ncurses.py:
```diff
@@ -0,0 +1,65 @@
+
+    VPkg: gnu, ncurses
+    """
+    regex = [r"ncurses ([0-9]+\.[0-9]+)"]
```
Should we extend this to expect a date at the end of the string, the way it appears in your compiled versions?  (We don't have to use the date, but it's there every time it's an addtional way to reduce false positives).

param211 replied:
We could do that, but there are some versions on rpmfind's page which don't have a date. Its probably safer to keep it like this.

terriko commented on cve_bin_tool/checkers/ncurses.py:
```diff
@@ -0,0 +1,65 @@
+    VPkg: gnu, ncurses
+    """
+    regex = [r"ncurses ([0-9]+\.[0-9]+)"]
+    regex1 = [r"tic-([0-9]+\.[0-9]+)"]
```
This signature makes me uncomfortable, beacuse it's a 3 letter differentiator and I feel like it's going to twig on something the ends in tic eventually.  Can you take a look at some other distros ncurses packages and see if that's the best we can do?

param211 replied:
@terriko  I added another regex, now version from regex1 and regex2 are used only if both of them are found. Possibly reducing the false positives

param211 commented:
Some new findings:
1. pkg from arch-linux contains "lib" folder (empty in rpmfind/fedora pkgs). This has binary files with lines matching the first regex ("ncurses x.x").
2. pkgs from rpmfind/fedora match regex1 and regex2. This is valid only for ncurses version >= 6.0
3. pkgs from rpmfind/fedora with version < 6.0 contain no strings which can identify a version number.

param211 commented:
@terriko I realised the way I was checking for regex1 and regex2 was wrong because they are found in different files (and regex_find gets lines from each file one at a time). This means we either have to find a different way to make sure that both regexes are present or we could switch to a single regex. I have added the regex "infocmp-x.x". "infocmp" seems long enough to avoid false positives.

note: the original regex "ncurses-x.x" remains intact. This is for the case when the original regex isn't found.
